### PR TITLE
fix distributed initialization

### PR DESCRIPTION
**Pull request by ZiyueHuang**
According to this line https://github.com/bytedance/byteps/blob/master/byteps/server/server.cc#L197, the server will initialize the stored parameter from the last init push which could be sent from an arbitrary worker.
cc @eric-haibin-lin

### CODE REVIEW

**Comment by vycezhong** [comment]
Actually, the server does not need to initialize. The `copy` here is useless. 
https://github.com/bytedance/byteps/blob/b8948f092796fa177c56a2a1595188dc7cce9999/byteps/server/server.cc#L209

This is because during each iteration, the server uses the first received data as the temporary buffer to sum. And after receiving gradients from all workers, the temporary buffer is copied to the `stored` and released. 
https://github.com/bytedance/byteps/blob/b8948f092796fa177c56a2a1595188dc7cce9999/byteps/server/server.cc#L227-L229

https://github.com/bytedance/byteps/blob/b8948f092796fa177c56a2a1595188dc7cce9999/byteps/server/server.cc#L274

**Comment by vycezhong** [comment]
I am afraid that it seems to matter for async mode...

File: byteps/mxnet/__init__.py [hunk 1]
```diff
@@ -221,8 +221,6 @@ def _init_params(self):
                 param_arrays = param._check_and_get(param._data, list)
                 idx = self._param2idx[param.name]
 
-                if rank() != self.root_rank:
-                    param_arrays[0].__imul__(0)
                 byteps_push_pull(param_arrays[0], version=0, priority=0,
                                  name="parameter_" + str(idx), is_average=False)
```
Comment (vycezhong) [inline byteps/mxnet/__init__.py:225]: May be we should use `is_average=True` here?

**Comment by ZiyueHuang** [comment]
For sync mode, this initialization is also necessary since it make sure that the parameters on each worker are the same at the begining. The current implementation is problematic since all the workers may acctually initialize the parameters to 0 (except when the root node sends the last init push), no matter what initializer the user chooses.

File: byteps/mxnet/__init__.py (hunk 1 above)
Comment (ZiyueHuang) [reply byteps/mxnet/__init__.py:225]: I think there is no need to average, since the server just selects one received tensor and copy it to the stored buffer.

Comment (vycezhong) [reply byteps/mxnet/__init__.py:225]: You're right.

**Comment by ZiyueHuang** [comment]
This problem also exists for the distributed gradient accumulation in the first round...

For the `push` phase in the first round of `byteps_push_pull`, the server will discard all the messages (except the last one) sent from all the workers. We should separate this step away from `byteps_push_pull` into a new api, say, `byteps_init`, and let `byteps_push_pull` only do the sum operation. Similar to `InitImpl` https://github.com/apache/incubator-mxnet/blob/master/src/kvstore/kvstore_dist.h#L193.

**Comment by vycezhong** [comment]
> This problem also exists for the distributed gradient accumulation in the first round...
> 
> For the `push` phase in the first round of `byteps_push_pull`, the server will discard all the messages (except the last one) sent from all the workers. We should separate this step away from `byteps_push_pull` into a new api, say, `byteps_init`, and let `byteps_push_pull` only do the sum operation. Similar to `InitImpl` https://github.com/apache/incubator-mxnet/blob/master/src/kvstore/kvstore_dist.h#L193.

I don't think there is such a problem in the first round of pushpull of gradients. The `stored` has been allocated and initialized during the pushpull of parameters. 

BTW, there is already an API called `bps.init()`.

**Comment by ZiyueHuang** [comment]
@vycezhong  The gradient and the weight for the same parameter are two different buffers. See https://github.com/bytedance/byteps/blob/master/byteps/mxnet/__init__.py#L203-L206

**Comment by vycezhong** [comment]
i see. you're right. I am sorry that i didn't notice parameter and gradient do not share the same buffer...

I think maybe we can reuse the parameter buffer?  for example, 

```python
        for i, param in enumerate(self._params):
            byteps_declare_tensor("data_" + str(i)) 
...

    def _init_params(self):
                ....
                byteps_push_pull(param_arrays[0], version=0, priority=0,
                                 name="data_" + str(idx), is_average=False)
```

in this way, there is only one collection of buffers. And `byteps_push_pull` in  `_init_params` will trigger initialization of the buffer. And then this buffer is used for gradients. I think it may be a solution.

**Comment by ZiyueHuang** [comment]
This might be a solution for training the model, but it is still weird that the first round of `byteps_push_pull` performs a random selection instead of the sum...

Furthermore, it may not be neccesary to maintain two buffers `update_buf_` and `store_ ` in https://github.com/bytedance/byteps/blob/master/byteps/server/server.h, since BytePS transmits gradients bidirectionally (both from worker to server and from server to worker). MXNet KVStore maintains these two buffers since MXNet pushes gradient from worker to server, then updates the weight on the server by the optimizer, then pulls weight from server to worker.

**Comment by eric-haibin-lin** [comment]
Init tensor comes from the root rank only: https://github.com/bytedance/byteps/blob/master/byteps/common/operations.cc#L331-L341

**Comment by vycezhong** [comment]
> Init tensor comes from the root rank only: https://github.com/bytedance/byteps/blob/master/byteps/common/operations.cc#L331-L341

Yes. But they use `rank()` (not `local_rank()`). For non-root nodes,  zeros are sent to servers and thus might be selected as the value of  `stored`. 
https://github.com/bytedance/byteps/blob/b8948f092796fa177c56a2a1595188dc7cce9999/byteps/mxnet/__init__.py#L224-L225

**Comment by bobzhuyb** [comment]
Is it causing any real problems? The initialized value is not used later, so we did not care to make it deterministic.

**Comment by vycezhong** [comment]
> Is it causing any real problems? The initialized value is not used later, so we did not care to make it deterministic.

Sorry for my careless. We found the logic is right.
